An "accurate and reliable" RDKitConverter

cbouy opened the pull request:
Hi @IAlibay @orbeckst @richardjgowers 

I made a few changes to the RDKitConverter so that it more accurately infers bond orders and charges in small molecules.
I tested this new version on the ChEMBL27 dataset and it managed to retrieve the original molecule or a resonance structure in 99.1% of the 1.7M compounds (the GSoC version was around 90%).
I can share the notebook that I used to run this ChEMBL benchmark, either in one of you other repos or I create a new one, as you prefer.

Changes made in this Pull Request:
 - Fixes monatomic ion charges that were always negative
 - Fixes more cases of conjugated systems
 - Fixes tetravalent phosphorus
 - Fixes some edge cases with sulfur and nitrogen
 - Modifies the bond inferring tests to try matching with resonance structures
 - Adds tests for ions, all standard amino-acids and all standard nucleic-acids
 - Adds tests on molecules encountered in the ChEMBL 27 dataset
 - Fixes atoms in the RDKit molecule that weren't always in the same order as in the atomgroup
 - Fixes atom name formatting when writing a PDB file with RDKit
 - Fixes sanitization errors that occurred when using `NoImplicit=False`
 - The `names` property is now available through `atom.GetProp("_MDAnalysis_name")`, while `atom.GetMonomerInfo().GetName()` returns the PDB-formatted version (with spaces).

All of these changes do not take "efficiency" in consideration, so it probably slows down the code a bit, but I think they are all necessary. If you identify things that could be optimized please let me know.

Their are additional changes that I could make that on one hand would slow down the code even more but on the other hand make it even more reliable. See #3339 for examples of failing molecules.

I could fix them by adding more SMIRKS/reaction SMARTS, but considering the fact that these patterns are rare and would slow down the conversion I don't think they are worth adding to the code.

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [ ] Issue raised/referenced?

pep8speaks commented:
Hello @cbouy! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`package/MDAnalysis/converters/RDKit.py`](https://github.com/MDAnalysis/mdanalysis/blob/f019d28350ef8adb9ba778b20dc1200e3fc4b426/package/MDAnalysis/converters/RDKit.py):

> [Line 81:80](https://github.com/MDAnalysis/mdanalysis/blob/f019d28350ef8adb9ba778b20dc1200e3fc4b426/package/MDAnalysis/converters/RDKit.py#L81): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 278:80](https://github.com/MDAnalysis/mdanalysis/blob/f019d28350ef8adb9ba778b20dc1200e3fc4b426/package/MDAnalysis/converters/RDKit.py#L278): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (129 > 79 characters)
> [Line 710:80](https://github.com/MDAnalysis/mdanalysis/blob/f019d28350ef8adb9ba778b20dc1200e3fc4b426/package/MDAnalysis/converters/RDKit.py#L710): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 79 characters)
> [Line 712:80](https://github.com/MDAnalysis/mdanalysis/blob/f019d28350ef8adb9ba778b20dc1200e3fc4b426/package/MDAnalysis/converters/RDKit.py#L712): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 79 characters)
> [Line 714:80](https://github.com/MDAnalysis/mdanalysis/blob/f019d28350ef8adb9ba778b20dc1200e3fc4b426/package/MDAnalysis/converters/RDKit.py#L714): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 79 characters)
> [Line 716:80](https://github.com/MDAnalysis/mdanalysis/blob/f019d28350ef8adb9ba778b20dc1200e3fc4b426/package/MDAnalysis/converters/RDKit.py#L716): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 79 characters)
> [Line 718:80](https://github.com/MDAnalysis/mdanalysis/blob/f019d28350ef8adb9ba778b20dc1200e3fc4b426/package/MDAnalysis/converters/RDKit.py#L718): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 79 characters)
> [Line 720:80](https://github.com/MDAnalysis/mdanalysis/blob/f019d28350ef8adb9ba778b20dc1200e3fc4b426/package/MDAnalysis/converters/RDKit.py#L720): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 79 characters)
> [Line 724:80](https://github.com/MDAnalysis/mdanalysis/blob/f019d28350ef8adb9ba778b20dc1200e3fc4b426/package/MDAnalysis/converters/RDKit.py#L724): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 79 characters)
> [Line 728:80](https://github.com/MDAnalysis/mdanalysis/blob/f019d28350ef8adb9ba778b20dc1200e3fc4b426/package/MDAnalysis/converters/RDKit.py#L728): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 79 characters)
> [Line 730:80](https://github.com/MDAnalysis/mdanalysis/blob/f019d28350ef8adb9ba778b20dc1200e3fc4b426/package/MDAnalysis/converters/RDKit.py#L730): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 79 characters)





* In the file [`testsuite/MDAnalysisTests/converters/test_rdkit.py`](https://github.com/MDAnalysis/mdanalysis/blob/f019d28350ef8adb9ba778b20dc1200e3fc4b426/testsuite/MDAnalysisTests/converters/test_rdkit.py):

> [Line 441:80](https://github.com/MDAnalysis/mdanalysis/blob/f019d28350ef8adb9ba778b20dc1200e3fc4b426/testsuite/MDAnalysisTests/converters/test_rdkit.py#L441): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 466:80](https://github.com/MDAnalysis/mdanalysis/blob/f019d28350ef8adb9ba778b20dc1200e3fc4b426/testsuite/MDAnalysisTests/converters/test_rdkit.py#L466): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)
> [Line 506:80](https://github.com/MDAnalysis/mdanalysis/blob/f019d28350ef8adb9ba778b20dc1200e3fc4b426/testsuite/MDAnalysisTests/converters/test_rdkit.py#L506): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)
> [Line 627:80](https://github.com/MDAnalysis/mdanalysis/blob/f019d28350ef8adb9ba778b20dc1200e3fc4b426/testsuite/MDAnalysisTests/converters/test_rdkit.py#L627): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 628:80](https://github.com/MDAnalysis/mdanalysis/blob/f019d28350ef8adb9ba778b20dc1200e3fc4b426/testsuite/MDAnalysisTests/converters/test_rdkit.py#L628): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 630:80](https://github.com/MDAnalysis/mdanalysis/blob/f019d28350ef8adb9ba778b20dc1200e3fc4b426/testsuite/MDAnalysisTests/converters/test_rdkit.py#L630): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)



##### Comment last updated at 2022-03-30 21:49:43 UTC

richardjgowers commented:
Hey @cbouy I think we're more interested in correctness rather than speed.  Speed is something we can tweak on later once we've established what the code should do.

lilyminium commented on package/MDAnalysis/coordinates/RDKit.py:
```diff
@@ -512,9 +512,25 @@ def _infer_bo_and_charges(mol):
     R-C(-O)-O the first oxygen read will receive a double bond and the other
     one will be charged. It will also affect more complex conjugated systems.
     """
+    atoms = sorted([a for a in mol.GetAtoms() if a.GetAtomicNum() > 1],
```
Will this cause issues for dummy atoms etc? e.g. ``rdkit.Chem.MolFromSmiles(r"C(=C(/C([2*:2]))N)(\[1*:1])"))``

cbouy replied:
I haven't considered dummy atoms when making the converter. It will very likely break the bond-order inferring since it needs the valence(s) of an atom to work properly but it's `-1` for dummies. Not sure if it would be useful or not

lilyminium commented on package/MDAnalysis/coordinates/RDKit.py:
```diff
@@ -512,9 +512,25 @@ def _infer_bo_and_charges(mol):
 
-    for atom in sorted(mol.GetAtoms(), reverse=True,
-                       key=lambda a: _get_nb_unpaired_electrons(a)[0]):
+    for atom in atoms:
```
Could you please add tests for case where an atom degree is 0?

lilyminium commented on package/MDAnalysis/coordinates/RDKit.py:
```diff
@@ -546,15 +562,16 @@ def _infer_bo_and_charges(mol):
-                # transform it to a negative charge
-                atom.SetFormalCharge(-nue)
+                    # go to next atom if one of the valences is complete
+                    nue = _get_nb_unpaired_electrons(atom)
+                    if any([n == 0 for n in nue]):
+                        break
```
Could you please add a test for this case where common_nue != 0?

lilyminium commented on package/MDAnalysis/coordinates/RDKit.py:
```diff
@@ -546,15 +562,16 @@ def _infer_bo_and_charges(mol):
+
+            # if atom valence is still not filled
+            nue = _get_nb_unpaired_electrons(atom)
+            if not any([n == 0 for n in nue]):
```
Could you add a test for this please?

lilyminium commented on package/MDAnalysis/coordinates/RDKit.py:
```diff
@@ -546,15 +562,16 @@ def _infer_bo_and_charges(mol):
+            nue = _get_nb_unpaired_electrons(atom)
+            if not any([n == 0 for n in nue]):
+                # transform nue to charge
+                atom.SetFormalCharge(-nue[0])
```
Could you add a test for this please?

lilyminium commented on package/MDAnalysis/coordinates/RDKit.py:
```diff
@@ -605,52 +622,53 @@ def _standardize_patterns(mol, max_iter=200):
 
+    # list of sanitized reactions
+    reactions = []
+    for name, rxn in [
```
You don't seem to use ``name``. Maybe they could be comments.

lilyminium commented on package/MDAnalysis/coordinates/RDKit.py:
```diff
@@ -680,27 +698,22 @@ def _run_reaction(reaction, reactant):
+        products = reaction.RunReactants((reactant,))
         # only keep the first product
         if products:
             product = products[0][0]
```
Is this guaranteed to be the "main" product?

lilyminium commented on package/MDAnalysis/coordinates/RDKit.py:
```diff
@@ -735,61 +749,83 @@ def _rebuild_conjugated_bonds(mol, max_iter=200):
-    elif n_matches % 2 == 0:
-        end_pattern = Chem.MolFromSmarts("[*-;!O]-[*+0]=[*+0]-[*-]")
-    else:
+    elif n_matches == 1:
```
Could you add a test for this please?

lilyminium commented on package/MDAnalysis/coordinates/RDKit.py:
```diff
@@ -735,61 +749,83 @@ def _rebuild_conjugated_bonds(mol, max_iter=200):
-        end_pattern = Chem.MolFromSmarts(
-            "[*-;!O]-[*+0]=[*+0]-[$([#7;X3;v3]),$([#6+0]=O)]")
+        end_pattern = odd_end_pattern
+    else:
```
Could you add a test for this please?

lilyminium commented on package/MDAnalysis/coordinates/RDKit.py:
```diff
@@ -735,61 +749,83 @@ def _rebuild_conjugated_bonds(mol, max_iter=200):
-                    if neighbor.GetAtomicNum() == 8 and bond.GetBondTypeAsDouble() == 2:
+                    if (neighbor.GetAtomicNum() == 8 and
+                        bond.GetBondTypeAsDouble() == 2):
                         bond.SetBondType(Chem.BondType.SINGLE)
```
Could you please add a comment saying that you're converting this double bond to a single bond because (I guess) it's an edge case in RDKit? Just to make it clear it's on purpose.

And same for the cases below.

lilyminium reviewed (changes requested):
Hi @cbouy, this looks awesome. Could you please add some tests for the lines that codecov points out, and address the comment by the PEP8 bot?

cbouy commented:
A couple more changes:
- atoms in the RDKit molecule are now in the same order as in the atomgroup: the BO inferring code could scramble atoms a bit but I found a convenient way to fix this. Now `mol.GetAtomWithIdx(i)` corresponds to the atom in `atomgroup[i]` as most people would expect
- fixing atom name formatting: this is only useful when converting the RDKit molecule to a PDB file, but I needed it (with the above fix) while playing with [py3Dmol](https://github.com/3dmol/3Dmol.js/tree/master/py3Dmol)
- sanitization is skipped if not possible: could happen when you have a degree 4 nitrogen atom but you were skipping BO/charge inferring so no charge was assigned to it, RDKit would complain about the uncharged v4 N during sanitization.
- a bit of performance boost: because I'm using reactions to fix some things when inferring BO, the properties that are stored in each atom had to be transferred "manually" after each reaction and it was painfully slow. I reorganized the code a bit to only transfer the `_MDAnalysis_index` property after each reaction, and do the complete transfer of properties once we have a "clean" molecule. Also the function to set properties has been changed to avoid running an `if... elif... else` for each atomic property.

orbeckst commented:
> Once all this is done I'll likely benchmark the BO inferring code on the latest ChEMBL, were should I upload the corresponding notebook?

You can always make a gist and share it. However, I think starting a repo (MDA_RDKit_benchmarks – come up with a better name) might actually be better as it has some room to grow and also gives you the opportunity to deposit data/derived data and eventually archive/publish data via Zenodo. The RDKit functionality is of interest to a sizable part of the community so starting to document and benchmark it will likely be a growing effort. 

As a member of the MDAnalysis org, you should be able to create a repository under MDAnalysis. Don't forget to add a license when you create the repo — GPLv3 is guaranteed to be a correct choice when you import MDAnalysis; a permissive BSD or MIT license will likely not be challenged by anyone.

orbeckst commented on package/MDAnalysis/converters/RDKit.py:
```diff
@@ -522,9 +563,25 @@ def _infer_bo_and_charges(mol):
+                   reverse=True,
+                   key=lambda a: _get_nb_unpaired_electrons(a)[0])
+
+    MONATOMIC_CATION_CHARGES = {
```
document data structure — is it atomic number: charge in e?

orbeckst commented on package/MDAnalysis/converters/RDKit.py:
```diff
@@ -733,6 +795,7 @@ def _rebuild_conjugated_bonds(mol, max_iter=200):
     Since ``anion-*=*`` is the same as ``*=*-anion`` in terms of SMARTS, we can
     control that we don't transform the same triplet of atoms back and forth by
     adding their indices to a list.
+    This function also handles conjugated systems with triple bonds.
```
Would it be more expensive to work on a copy of `mol` and return a new `mol`?

cbouy replied:
It should be fine but I fail to see the point of working on a copy

orbeckst reviewed (commented):
I only had a very shallow look at things. On the documentation side this is looking pretty good. Perhaps some general documentation at the top of the module that outlines your overall approach, high level ideas, and algorithms, like a Methods section in a paper.

I like the comments in the helper functions and that they appear in the docs. I like comments in the code but as @lilyminium pointed out somewhere, add more, especially to explain _why_ something is being done. 

Overall, excellent addition.

cbouy commented:
I've added the link to my RDKit UGM presentation (both video and slides) so that there's an alternative and more visual explanation to how the converter works.

cbouy commented:
For tackling the remaining known failures (molecules containing `R=[N+](-[O-])-R`, `R-[N+]#N`, `R-N=[N+]=[N-]`, `R-[O+]=R`, `R-[N+]#[C-]`, `R-[N-]-R`) should I stick to this PR or open a new one?
I feel like this PR is already getting pretty lengthy and sprawling (my fault) so I'd rather get this essential part merged, then run and publish the benchmark, and see later for the rest as I'm not sure how available I'll be in the near future

lilyminium commented:
@cbouy yeah I'd make a new branch/PR.

orbeckst commented:
Could you please open an issue for the remaining known failures https://github.com/MDAnalysis/mdanalysis/pull/3044#issuecomment-848331158, @cbouy ? Comments get lost, issues can get addressed (eventually). They also provide documentation for what we _know_ is not working properly and can serve as a rallying point for users who experience problems. Thanks.

orbeckst commented:
An additional way to document the failing patterns would be to include tests that XFAIL. Then in the next step, one "just" has to add code to make these tests pass.

IAlibay commented:
@cbouy @orbeckst @lilyminium what do we want to do with this re: 2.0? I'd like to get 2.0 out ASAP, can this be pushed to 2.1 instead? We could include a warning in the docstring of the converter that would go somewhere along the lines of "this is experimental, improved accuracy is coming in a future release"?

orbeckst commented:
An "experimental" warning should be sufficient, perhaps with a link to this issue.

IAlibay commented:
@cbouy @lilyminium @orbeckst what needs to be done here to get this merged on time for 2.1.0?

If any @MDAnalysis/coredevs want to take over pushing this along it would be great.

cbouy commented:
@IAlibay I think I needed to find one or two more example molecules to have perfect coverage.
Is there a deadline for 2.1.0?

IAlibay commented:
> @IAlibay I think I needed to find one or two more example molecules to have perfect coverage. Is there a deadline for 2.1.0?

It's pretty much "as soon as" - I think the original deadline was 2 months ago 😅

cbouy commented:
All right, I'll give it a shot in the coming days then

cbouy commented:
Aaaaand... I think that's it! Ready for review
![salt bae](https://media2.giphy.com/media/l4Jz3a8jO92crUlWM/giphy.gif)

IAlibay commented:
@lilyminium @orbeckst can you check that your comments have been addressed here?

orbeckst reviewed (changes requested):
I don't have the expertise to give this a thorough review so I am commenting in general.

- Please add a CHANGELOG entry.
- Are there still cases that fail? If so, can you document these cases somehow — either mention in the docs or include XFAIL tests?

lilyminium commented on package/MDAnalysis/converters/RDKit.py:
```diff
@@ -325,9 +330,9 @@ def convert(self, obj, cache=True, NoImplicit=True, max_iter=200,
                 conf = Chem.Conformer(mol.GetNumAtoms())
                 for atom in mol.GetAtoms():
-                    idx = atom.GetIntProp("_MDAnalysis_index")
+                    idx = atom.GetIdx()
```
Just commenting to think out loud -- this is fine because `_MDAnalysis_index` is set sequentially in `atomgroup_to_mol`.

lilyminium commented on package/MDAnalysis/converters/RDKit.py:
```diff
@@ -435,9 +440,16 @@ def atomgroup_to_mol(ag, NoImplicit=True, max_iter=200, force=False):
         # infer bond orders and formal charges from the connectivity
         _infer_bo_and_charges(mol)
         mol = _standardize_patterns(mol, max_iter)
+        # reorder atoms to match MDAnalysis
+        order = np.argsort([atom.GetIntProp("_MDAnalysis_index")
+                            for atom in mol.GetAtoms()])
+        mol = Chem.RenumberAtoms(mol, order.astype(int).tolist())
```
Question: Is this necessary? Shouldn't _Mdanalysis_index have been set sequentially above in the `enumerate` call?

lilyminium replied:
Also, `MDAnalysis_index` is a bit ambiguous in the sense that I don't really know if it's the `atom.index` property or index in that particular atom group. Possibly out of scope for this PR.

lilyminium replied:
Oh wait, I'm guessing this is because the RDKit reactions will have messed up the order. Would you mind adding a comment to that effect?

cbouy replied:
> Also, `MDAnalysis_index` is a bit ambiguous in the sense that I don't really know if it's the `atom.index` property or index in that particular atom group.

It's the index in that specific atomgroup, I'll add a note on that.


> Oh wait, I'm guessing this is because the RDKit reactions will have messed up the order. Would you mind adding a comment to that effect?

Yup!

lilyminium commented on package/MDAnalysis/converters/RDKit.py:
```diff
@@ -475,30 +487,68 @@ def _add_mda_attr_to_rdkit(attr, value, mi):
-            symbol, number = name
-            if len(number) > 2 and len(symbol) == 1:
-                value = "{}{}".format(symbol, number)
+            # only two characters
+            if len(value) == 2:
+                value = f" {value} "
+            # more than two --> cut to 4 characters
             else:
-                value = "{:>2.2}{:<2.2}".format(symbol, number)
+                symbol, number = name
+                value = f"{symbol + number:>4.4}"
+        # no number in the name
         else:
-            # no number in the name
-            value = " {:<}".format(name[0])
+            value = f" {name[0]:<3.3}"
```
I'm guessing this is for RDKit's PDB writer, but it's IMO suboptimal to edit the names like this. It'd be nice to set another property with `SetProp` without this character limit, if this property is used for conversion from RDKit into MDAnalysis as well.

lilyminium replied:
I might also have missed sometihng but I don't fully understand the left 3-column justification if there's no number in the name -- is this an RDKit quirk? Otherwise you could maybe just use the `center` function: `f"{value.center(4):4.4}"`

cbouy replied:
I think a better approach would be to just reuse the `PDBWriter._deduce_PDB_atom_name` method instead of reinventing the wheel. It requires residue names to work though, so I'll need to refactor my code a bit but it's for the best.

Should I add a `_MDAnalysis_name` property to keep the original name just in case?

> I don't fully understand the left 3-column justification if there's no number in the name -- is this an RDKit quirk? 

There's no specific PDB or RDKit rule for atom names without numbers, it's just that I originally organized my code in "names with numbers" vs "names without numbers" because I thought that was a more pragmatic approach, although there's likely a much better way to do it.

> Otherwise you could maybe just use the `center` function: `f"{value.center(4):4.4}"`

`OXT` would become `OXT␣` instead of `␣OXT`

lilyminium commented on package/MDAnalysis/converters/RDKit.py:
```diff
@@ -693,27 +774,30 @@ def _run_reaction(reaction, reactant):
+    for n in range(reactant.GetNumAtoms()):
+        reactant.UpdatePropertyCache(strict=False)
+        Chem.Kekulize(reactant)
+        products = reaction.RunReactants((reactant,))
```
Would this be a good spot for `RunReactantInPlace`? Would that then save the property updating?

cbouy replied:
Yes, although it's been introduced in RDKit v2021.09.1 so I'm not sure if it's a good idea to implement it in the code right now.  
It will very likely speed things up a ton (and simplify the codebase), but we can focus on optimization later I think ?

lilyminium commented on package/MDAnalysis/core/selection.py:
```diff
@@ -679,12 +679,10 @@ def _apply(self, group):
-            mol.GetAtomWithIdx(idx).GetIntProp("_MDAnalysis_index")
-            for match in matches for idx in match]
+        # flatten all matches and remove duplicated indices
+        indices = np.unique([idx for match in matches for idx in match])
         # create boolean mask for atoms based on index
-        mask = np.in1d(range(group.n_atoms), np.unique(indices))
+        mask = np.in1d(range(group.n_atoms), indices)
```
LGTM. this works because the entire group is converted and therefore RDKit index should == MDAnalysis_index (relative to `group`).

lilyminium commented on package/MDAnalysis/topology/guessers.py:
```diff
@@ -518,7 +516,6 @@ def guess_gasteiger_charges(atomgroup):
-    return np.array([atom.GetDoubleProp("_GasteigerCharge") for atom in atoms],
+    return np.array([atom.GetDoubleProp("_GasteigerCharge")
+                     for atom in mol.GetAtoms()],
                     dtype=np.float32)
```
LGTM

lilyminium commented on testsuite/MDAnalysisTests/converters/test_rdkit.py:
```diff
@@ -204,13 +189,12 @@ def test_monomer_info(self, pdb, sel_str, atom_index):
 
         for mda_attr, rd_attr in RDATTRIBUTES.items():
             rd_value = getattr(mi, "Get%s" % rd_attr)()
             if hasattr(sel, mda_attr):
-                mda_value = getattr(sel, mda_attr)[mda_index]
+                mda_value = getattr(sel, mda_attr)[atom_index]
```
I find this somewhat hard to read, to be honest. I don't think this class is inherited, so could you please just type out the attributes you expect the AtomGroup to have?

lilyminium replied:
i.e.

```
assert sel.altLocs[atom_index] == monomer_info.GetAltLoc()
assert sel.names[atom_index] == monomer_info.GetName().strip()
```

lilyminium reviewed (changes requested):
Sorry, this is a partial review. I think everything looks good to me so far, but I need a bit of time to wrap my head around the chemoinformatics part.

lilyminium commented on package/MDAnalysis/converters/RDKit.py:
```diff
@@ -618,60 +690,69 @@ def _standardize_patterns(mol, max_iter=200):
+    # reassign all properties
+    _transfer_properties(mol, newmol)
+
+    # fix bonds with "crossed" stereo
+    for bond in newmol.GetBonds():
+        if bond.GetStereo() == Chem.BondStereo.STEREOANY:
+            bond.SetStereo(Chem.BondStereo.STEREONONE)
+
+    return newmol
```
Is this because the reactions might have changed altered the stereochemistry? Do you mind commenting that? What would be the harm in leaving them as Stereoany?

cbouy replied:
I think this isn't needed anymore, I introduced it when the converter was just reading the topology (so very early in the GSoC project). Since then I added parsing coordinates and assigning stereo from the coordinates so I'll just remove it and run the tests.

lilyminium commented on testsuite/MDAnalysisTests/converters/test_rdkit.py:
```diff
@@ -397,9 +378,36 @@ def test_cache(self):
 @requires_rdkit
 class TestRDKitFunctions(object):
+
+    def mol_to_template(self, mol):
+        """Remove bond orders and charges from molecule"""
+        template = Chem.AddHs(mol)
+        for atom in template.GetAtoms():
+            atom.SetIsAromatic(False)
+            atom.SetFormalCharge(0)
+            atom.SetNoImplicit(True)
+        for bond in template.GetBonds():
+            bond.SetIsAromatic(False)
+            bond.SetBondType(Chem.BondType.SINGLE)
+        template.UpdatePropertyCache(strict=False)
+        return template
```
Does template have a special meaning here? If not, could you please rename this to something like `remove_bond_orders_and_charges` or something a bit more descriptive for the tests?

lilyminium commented on testsuite/MDAnalysisTests/converters/test_rdkit.py:
```diff
@@ -484,82 +498,188 @@ def test_set_atom_property(self, attr, value, getter):
                           match="reasonable number of iterations"):
             _rebuild_conjugated_bonds(mol, 2)
+
+    @pytest.mark.parametrize("smi", [
+        "[Li+]", "[Na+]", "[K+]", "[Rb+]", "[Ag+]", "[Cs+]",
+        "[Mg+2]", "[Ca+2]", "[Cu+2]", "[Zn+2]", "[Sr+2]", "[Ba+2]",
+        "[Al+3]", "[Fe+2]",
+        "[Cl-]",
+        "[O-2]",
+        "[Na+].[Cl-]",
+    ])
+    def test_ions(self, smi):
+        ref = Chem.MolFromSmiles(smi)
+        mol = self.mol_to_template(ref)
+        _infer_bo_and_charges(mol)
+        mol = _standardize_patterns(mol)
+        Chem.SanitizeMol(mol)
+        assert mol.HasSubstructMatch(ref) and ref.HasSubstructMatch(mol)
```
Test LGTM

lilyminium commented on testsuite/MDAnalysisTests/converters/test_rdkit.py:
```diff
@@ -484,82 +498,188 @@ def test_set_atom_property(self, attr, value, getter):
         for a in template.GetAtoms():
             smi = Chem.MolToSmiles(template, rootedAtAtom=a.GetIdx())
             m = Chem.MolFromSmiles(smi, sanitize=False)
             for atom in m.GetAtoms():
                 atom.SetFormalCharge(0)
+                atom.SetIsAromatic(False)
                 atom.SetNoImplicit(True)
```
Doesn't `mol_to_template` already do this?

cbouy replied:
you're right, only `atom.SetNoImplicit(True)` is needed

lilyminium commented on testsuite/MDAnalysisTests/converters/test_rdkit.py:
```diff
@@ -484,82 +498,188 @@ def test_set_atom_property(self, attr, value, getter):
+                atom.SetIsAromatic(False)
                 atom.SetNoImplicit(True)
             m.UpdatePropertyCache(strict=False)
             _infer_bo_and_charges(m)
             m = _standardize_patterns(m)
```
This shows up in a few different places. Would it be easier to test and use if there was an overall wrapping function e.g. below (probably with a better name)?

```
def _cheminformagic(rdmol):
    _infer_bo_and_charges(rdmol)
    _standardize_patterns(rdmol)
```

lilyminium commented on testsuite/MDAnalysisTests/converters/test_rdkit.py:
```diff
@@ -484,82 +498,188 @@ def test_set_atom_property(self, attr, value, getter):
-                m), (f"(input) {Chem.MolToSmiles(ref)} != "
-                     f"{Chem.MolToSmiles(m)} (output) root atom {a.GetIdx()}")
+            match = m.HasSubstructMatch(ref) and ref.HasSubstructMatch(m)
+            if not match:
+                # try resonance structures for charged conjugated systems
+                for mol in Chem.ResonanceMolSupplier(m, maxStructs=20):
+                    match = (mol.HasSubstructMatch(ref) and
+                             ref.HasSubstructMatch(mol))
+                    if match:
+                        break
```
Would it be difficult to break out the test cases where you run into resonance problems, and explicitly test that resonance structure, in a separate test? i.e. `def test_order_independent_resonance`

cbouy replied:
I think this would mostly concern the molecules with a charge next to a conjugated system, which is a good portion of the tests here, but I'm not sure how separating the tests is going to make things more readable. It would IMO make things more difficult when/if we later add other tests.
I think a good option is to reorganize the test code with more functions with descriptive names, as you suggested in other comments

lilyminium commented on testsuite/MDAnalysisTests/converters/test_rdkit.py:
```diff
@@ -484,82 +498,188 @@ def test_set_atom_property(self, attr, value, getter):
-            bond.SetIsAromatic(False)
-            bond.SetBondType(Chem.BondType.SINGLE)
-
+        if input_type == "smi":
+            ref = Chem.MolFromSmiles(input_str)
+        elif input_type == "na":
+            ref = Chem.MolFromSequence(input_str, flavor=9)
+        else:
+            ref = Chem.MolFromSequence(input_str)
+        template = self.mol_to_template(ref)
         # go through each possible starting atom
         for a in template.GetAtoms():
```
This test has a lot of logic in it, to the point where the test itself almost needs a test. Could you turn the `for` loop into a separate function, e.g. `_assert_order_independent`, and separate the smi/na/aa molecules into separate tests?

lilyminium replied:
Or you could just use smiles for the acids, unless testing the RDKit sequence is important / it's likely to change internally in coming releases.

lilyminium commented on testsuite/MDAnalysisTests/converters/test_rdkit.py:
```diff
@@ -484,82 +498,188 @@ def test_set_atom_property(self, attr, value, getter):
+        for a in newmol.GetAtoms():
+            ix = a.GetIntProp("_MDAnalysis_index")
+            new[ix] = {"_MDAnalysis_index": ix,
+                       "dummy": a.GetProp("dummy")}
```
```suggestion
                       "dummy": a.GetProp(f"foo_{ix}")}
```
Otherwise I'm not sure this tests much.

cbouy replied:
Nope the test is fine, the name of the property is "dummy" and its value should be "foo_{ix}", so to retrieve the value we just call `atom.GetProp("dummy")`

lilyminium commented on testsuite/MDAnalysisTests/converters/test_rdkit.py:
```diff
@@ -471,8 +479,14 @@ def test_standardize_patterns(self, smi, out):
 
     @pytest.mark.parametrize("attr, value, getter", [
         ("index", 42, "GetIntProp"),
+        ("index", np.int8(42), "GetIntProp"),
+        ("index", np.int16(42), "GetIntProp"),
         ("index", np.int32(42), "GetIntProp"),
         ("index", np.int64(42), "GetIntProp"),
+        ("index", np.uint8(42), "GetIntProp"),
+        ("index", np.uint16(42), "GetIntProp"),
+        ("index", np.uint32(42), "GetIntProp"),
+        ("index", np.uint64(42), "GetIntProp"),
```
At this rate it might be worth just writing a separate test for getting `index` specifically.

lilyminium commented on testsuite/MDAnalysisTests/converters/test_rdkit.py:
```diff
@@ -484,82 +498,188 @@ def test_set_atom_property(self, attr, value, getter):
-            assert m.HasSubstructMatch(ref) and ref.HasSubstructMatch(
-                m), (f"(input) {Chem.MolToSmiles(ref)} != "
-                     f"{Chem.MolToSmiles(m)} (output) root atom {a.GetIdx()}")
+            match = m.HasSubstructMatch(ref) and ref.HasSubstructMatch(m)
```
This is short, but it might still be nice to make it a function with a descriptive name, like `assert_is_isomorphic` or something. It took me a minute to work out that this is the actual test where you put in a template and get out the right structure. A docstring would be extremely helpful for this test too.

lilyminium commented on package/MDAnalysis/converters/RDKit.py:
```diff
@@ -618,60 +690,69 @@ def _standardize_patterns(mol, max_iter=200):
 
+    # list of sanitized reactions
+    reactions = []
+    for rxn in [
+        "[C-;X2;H0:1]=[O:2]>>[C+0:1]=[O:2]",  # Cterm
+        "[N-;X2;H1;$(N-[*^3]):1]>>[N+0:1]",  # Nterm
+        "[#6-:1]-[#6:2]=[O:3]>>[#6+0:1]=[#6:2]-[O-:3]",  # keto-enolate
+        "[C-;v3:1]-[#7+0;v3;H2:2]>>[#6+0:1]=[#7+:2]",  # ARG
+        "[#6+0;H0:1]=[#6+0:2]-[#7;X3:3]-[#6-;X3:4]"
+        ">>[#6:1]=[#6:2]-[#7+:3]=[#6+0:4]",  # HIP
+        "[S;D4;!v6:1]-[*-:2]>>[S;v6:1]=[*+0:2]",  # sulfone
+        "[#7+0;X3:1]-[*-:2]>>[#7+:1]=[*+0:2]",  # charged-N
+    ]:
```
<img width="459" alt="Screen Shot 2022-03-02 at 2 08 46 am" src="https://user-images.githubusercontent.com/31115101/156194338-fbe496ad-2e71-4bdf-b082-651b4083389a.png">

I think this all LGTM, although I needed to convince myself that the `[C-]-N` to `[C]=[N+]` transformations generally applicable.

cbouy commented:
Thanks a ton for reviewing this again @lilyminium and @orbeckst 👍
I'll give it a go in the coming weeks

cbouy commented:
Okay I should be done with this one now!

I realized that there was a better way to test if the mol matches one of the resonance structure, so I implemented this change.

I added the XFAIL tests as @orbeckst suggested

Thanks again for reviewing this!!

cbouy commented:
I'm also preparing the benchmark on ChEMBL for the converter, should I create a new `RDKitConverter-benchmark` repository inside MDAnalysis then?

orbeckst commented:
Yes, create the repo under the MDAnalysis org. (I _think_ you have permission to do so but if not say something.)

orbeckst commented on package/MDAnalysis/converters/RDKit.py:
```diff
@@ -53,7 +53,7 @@
     infers the structures of approximately 90% of the `ChEMBL27`_ dataset.
```
Should this be updated to 
> 99.1% of the 1.7M compounds in the CHEMBL27 dataset

orbeckst replied:
Is the whole warning still necessary or can it become a Note or just text?

cbouy replied:
> Should this be updated to
> 
> > 99.1% of the 1.7M compounds in the CHEMBL27 dataset

Yes! good catch

> Is the whole warning still necessary or can it become a Note or just text?

I think it's necessary to keep the warning, just to catch people's attention that it's not a perfect solution and there are a few cases that are known to fail. I'll update this part a bit though.

orbeckst commented on package/CHANGELOG:
```diff
@@ -13,13 +13,23 @@ The rules for this file:
  * 2.2.0
 
 Fixes
+  * When converting an AtomGroup to an RDKit molecule (PR #3044):
```
Does this fix #3339 ? If so, the issue number should appear here. Even if it's only a partial fix — issue numbers are generally more important than PRs although we typically cite _both_ when possible.

cbouy replied:
No, #3339 keeps track of the cases that this PR cannot fix

orbeckst commented on package/MDAnalysis/converters/RDKit.py:
```diff
@@ -53,7 +53,7 @@
     infers the structures of approximately 90% of the `ChEMBL27`_ dataset.
     Work is currently ongoing on further improving this and updates to the
     converter are expected in future releases of MDAnalysis.
-    Please see `Pull Request #3044`_ for more details.
+    Please see `Pull Request #3044`_ and `Issue #3339`_ for more details.
```
I think the link to the PR can be removed. Better write some text that describes the current capabilities and lists the remaining known limitations. If there's an issue for these limitations, link to the issue.

orbeckst commented on package/MDAnalysis/converters/RDKit.py:
```diff
@@ -262,9 +267,32 @@ class RDKitConverter(base.ConverterBase):
     .. versionadded:: 2.0.0
 
+    .. versionchanged:: 2.2.0
+        See the full changelog for more details:
```
Please don't reference the CHANGELOG. The entry should be self-contained and provide the most important points. I also don't think that item lists properly format. They seem to be typeset separately. Just write prose over multiple lines. 

I don't think that the SanitizationError fix requires a "versionchanged" entry — this sounds like a bug fix to me. Mostly the versionchanged tells users if they need to be aware of changes in usage or output. (Although, if in dount, provide rather more information than less ;-) ).

orbeckst commented on package/MDAnalysis/converters/RDKit.py:
```diff
@@ -277,8 +305,8 @@ def convert(self, obj, cache=True, NoImplicit=True, max_iter=200,
         -----------
-        obj : :class:`~MDAnalysis.core.groups.AtomGroup` or :class:`~MDAnalysis.core.universe.Universe`
-
+        obj : :class:`~MDAnalysis.core.groups.AtomGroup` or
+            :class:`~MDAnalysis.core.universe.Universe`
```
has to be on one line to format correctly, see https://mdanalysis--3044.org.readthedocs.build/en/3044/documentation_pages/converters/RDKit.html#MDAnalysis.converters.RDKit.RDKitConverter.convert

orbeckst replied:
It might work if you just say
```
obj : `AtomGroup` or `Universe`
```
Try it and see if it links correctly.

orbeckst replied:
The back-tic notation did not automagically generate links so your previous solution was better. I'd revert to it.

orbeckst commented on testsuite/MDAnalysisTests/converters/test_rdkit.py:
```diff
@@ -529,37 +581,128 @@ def test_reassign_props_after_reaction(self, rdmol, product, name):
+    @pytest.mark.parametrize("smi", [
+        "[Li+]", "[Na+]", "[K+]", "[Rb+]", "[Ag+]", "[Cs+]",
+        "[Mg+2]", "[Ca+2]", "[Cu+2]", "[Zn+2]", "[Sr+2]", "[Ba+2]",
+        "[Al+3]", "[Fe+2]",
```
What happens for `"[Fe+3]"` or `"[Cu+1]"`?

orbeckst replied:
If nothing else, document known failure cases or make choices explicit in the docs. E.g., it's fine to say "Any iron ion will be guessed to be Fe3+" etc. We just need to be clear about what happens.

cbouy replied:
> What happens for `"[Fe+3]"` or `"[Cu+1]"`?

Those will always be converted to Fe2+ and Cu2+ with the code in this PR. We could solve this by looking at naming conventions for these ions but that's for another PR I guess. I'll add it to issue #3339 though

orbeckst reviewed (changes requested):
Looks very impressive. I could only comment on documentation aspects and one question about cations.

lilyminium commented:
Aside from @orbeckst's comments the code looks great -- thank you for putting up with the super slow reviews! Agreed with Oliver, the guessing code is very impressive and it'll be a fantastic solution for so many people trying to create molecular graphs out of lossy input like xyz files.

orbeckst commented:
@cbouy I don't know how busy you are at the moment but if you could have a look at my minor comments then we could merge this big PR really soon™.

cbouy commented:
@orbeckst I'm in the middle of moving out but I should be able to find some time after Monday! 
Thanks again to both of you!

orbeckst reviewed (approved):
Thanks for answering my questions — looks good.

(There's one change that I would revert so that links are generated in the docs but that's not blocking.)

Great contribution!!

orbeckst commented:
@lilyminium or @IAlibay can you please be in charge of eventually merging? Thanks!

cbouy commented:
Reverted the docstring thingy (and fixed a typo)!

Thanks a ton @orbeckst @lilyminium and @IAlibay for your help!!

![supernatural](https://user-images.githubusercontent.com/27850535/160938081-50164d9d-afa8-49af-aecd-ea1322f6e61d.gif)

orbeckst commented:
@lilyminium @IAlibay is something holding up the merge? I just don't know, otherwise I'd press the shiny green button myself.

(Sorry to be onerous, I just want to see this code in before anything else happened.)

IAlibay commented:
Sorry just really busy atm, I'll have a quick glance through over the weekend and then merge.

lilyminium commented:
No -- just had another quick glance and everything still LGTM! @IAlibay did you still want to have a look? Otherwise I can merge.

IAlibay commented:
> No -- just had another quick glance and everything still LGTM! @IAlibay did you still want to have a look? Otherwise I can merge.

Go for it!

orbeckst commented:
Hooray!!!!
